src/Conversion/ONNXToKrnl/Tensor/Squeeze.cpp

```diff
@@ -0,0 +1,114 @@
+//===--------------- Squeeze.cpp - Lowering Squeeze Op --------------------===//
+//
+// Copyright 2019 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Squeeze Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+
+using namespace mlir;
+
+struct ONNXSqueezeOpLowering : public ConversionPattern {
+  ONNXSqueezeOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXSqueezeOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXSqueezeOpOperandAdaptor operandAdaptor(operands);
+    auto loc = op->getLoc();
+    auto memRefType = convertToMemRefType(*op->result_type_begin());
+    auto memRefShape = memRefType.getShape();
+    auto elementSizeInBytes = getMemRefEltSizeInBytes(memRefType);
+    Value data = operandAdaptor.data();
+
+    // If only this op uses the input, then simply construct from the input a
+    // MemRef of a new type, and return the new MemRef.
+    //
+    // Use `op->getOperands()` instead of `operands` in order to get the number
+    // input uses.
+    //
+    // KrnlGetRef requires static shapes for the output, so perform a sanity
+    // check for this.
+    if (op->getOperands()[0].hasOneUse() &&
```

Comment:
Not sure if using a Kernel GetRef is the right operation here... as we now think of get refs as a way to allocate memory from a byte memory pool.

%comp = ONNX_op(...)
%squeezed = ONNX_squeeze(%comp)...

if there are no other uses of %comp, we can definitely just change the dimensions of %comp itself... just using memrefs below with no "1" in the dimensions.

But if there was another op that used %comp but unsqueezed, its the same data expect that the memrefs would not be updated... so its almost a KRNL->KRNL optimization where the squeezed is removed by changing the dimensionality of the memrefs afterwards... what do you think?

Reply:
Actually, the problem was I could not change a memref's shape in place, so I used krnl.getref to create a new view by removing "1"s in the dimensions. I tried to use std.MemRefCast, but it does not support a cast with different ranks like:
```mlir
%0 = memref_cast %X : memref<1x10x10xf32> to memref<10x10xf32>
```
However, we can write:
```mlir
%0 = memref_cast %X : memref<1x10x10xf32> to memref<*xf32>
%1 = memref_cast %0 : memref<*xf32> to memref<10x10xf32>
```
but it does allocate a temporary memory as follows:
```mlir
   // For the first instruction '%0 = ...' 
    %0 = llvm.mlir.undef : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %1 = llvm.insertvalue %arg0, %0[0] : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %2 = llvm.insertvalue %arg1, %1[1] : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %3 = llvm.insertvalue %arg2, %2[2] : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %4 = llvm.insertvalue %arg3, %3[3, 0] : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %5 = llvm.insertvalue %arg6, %4[4, 0] : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %6 = llvm.insertvalue %arg4, %5[3, 1] : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %7 = llvm.insertvalue %arg7, %6[4, 1] : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %8 = llvm.insertvalue %arg5, %7[3, 2] : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %9 = llvm.insertvalue %arg8, %8[4, 2] : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }">
    %10 = llvm.mlir.constant(1 : index) : !llvm.i64
    %11 = llvm.alloca %10 x !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }"> : (!llvm.i64) -> !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }*">
    llvm.store %9, %11 : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }*">
    %12 = llvm.bitcast %11 : !llvm<"{ float*, float*, i64, [3 x i64], [3 x i64] }*"> to !llvm<"i8*">

    // For the second instruction '%1 = ...'
    %13 = llvm.mlir.constant(3 : i64) : !llvm.i64
    %14 = llvm.mlir.undef : !llvm<"{ i64, i8* }">
    %15 = llvm.insertvalue %13, %14[0] : !llvm<"{ i64, i8* }">
    %16 = llvm.insertvalue %12, %15[1] : !llvm<"{ i64, i8* }">
    %17 = llvm.extractvalue %16[1] : !llvm<"{ i64, i8* }">
    %18 = llvm.bitcast %17 : !llvm<"i8*"> to !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }*">
    %19 = llvm.load %18 : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }*">
    llvm.return %19 : !llvm<"{ float*, float*, i64, [2 x i64], [2 x i64] }">
```

If krnl.getref is inappropriate, I suggest to remove such a "squeeze in place" optimization at this lowering level, and we will go back to this optimization when we find a better solution. Doing the optimization at Krnl->Krnl as you said is also a good way, I think.

Reply:
I also support removing such a "squeeze in place" optimization because the assumption underlying this check:
```
if (op->getOperands()[0].hasOneUse() &&
```
is that the memory references referred to by distinct MemRef SSA values do not overlap, which is an invariant this exact lowering pass would break. In other words, if we allow MemRefs to alias one another, as implied by this "squeeze in place" optimization, then checks like `op->getOperands()[0].hasOneUse()` would not be sufficient to determine the validity of optimization, since aliases of `op->getOperands()[0]` could nevertheless be used by other operations while being undetected by this check.

In general, I think the deviation from the "no-alias" principle of MemRefs should be postponed as late as possible.

Reply:
ok, that make sense